### PR TITLE
More account methods

### PR DESCRIPTION
**Pull request by BerryLiu2002**
### Summary
Added methods
get_assignment_submissions
get_assignments
get_assignment_submission

### Details
_Add more context to describe the  changes..._

### Checks
- [x] Tested changes
- [ ] Attached Logs

### Team to Review
_Mention the name of the team to review the PR_

### Reference to the issue
_Mention the issue ID or resources_

### CODE REVIEW

File: src/gradescopeapi/_classes/_account.py [hunk 1]
```diff
@@ -70,3 +77,124 @@ def get_courses(self) -> dict:
         courses["student"] = student_courses
 
         return courses
+
+    def get_assignments(self, course_id: int) -> List[Dict[str, str]]:
```
Comment (calvinatian) [inline src/gradescopeapi/_classes/_account.py:81]:
```suggestion
    def get_assignments(self, course_id: str) -> List[Dict[str, str]]:
```
I think we are keeping `course_id` as strings for now. I think keeping them as `ints` would also be OK but we would need to change all places we are using the `course_id` to `int`

File: src/gradescopeapi/_classes/_account.py [hunk 2]
```diff
@@ -70,3 +77,124 @@ def get_courses(self) -> dict:
...
+                "You must be logged in to access this page.": if no user is logged in
+                "Page not Found": When link is invalid: change in url, invalid course_if or assignment id
+                "Image only submissions not yet supported": assignment is image submission only, which is not yet supported
+        NOTE:
+        1. Image submissions not supports, need to find an endpoint to retrieve image pdfs
+        2. Not recommended for use, since this makes a GET request for every submission -> very slow!
+        3. so far only accessible for teachers, not for students to get submissions to an assignment
+        """
+        ASSIGNMENT_ENDPOINT = f"https://www.gradescope.com/courses/{course_id}/assignments/{assignment_id}"
+        ASSIGNMENT_SUBMISSIONS_ENDPOINT = f"{ASSIGNMENT_ENDPOINT}/review_grades"
```
Comment (calvinatian) [inline src/gradescopeapi/_classes/_account.py:154]:
```suggestion
        ASSIGNMENT_SUBMISSIONS_ENDPOINT = f"https://www.gradescope.com/courses/{course_id}/assignments/{assignment_id}/review_grades"
```

File: src/gradescopeapi/_classes/_account.py [hunk 3]
```diff
@@ -70,3 +77,124 @@ def get_courses(self) -> dict:
...
+            raise Exception("One or more invalid parameters")
+        session = self.session
+        submissions_resp = check_page_auth(session, ASSIGNMENT_SUBMISSIONS_ENDPOINT)
+        submissions_soup = BeautifulSoup(submissions_resp.text, "html.parser")
+        submissions_a_tags = submissions_soup.select("td.table--primaryLink a")
+        submission_ids = [
+            a_tag.attrs.get("href").split("/")[-1] for a_tag in submissions_a_tags
+        ]
+        submission_links = {}
+        for submission_id in submission_ids:  # doesn't support image submissions yet
```
Comment (calvinatian) [inline src/gradescopeapi/_classes/_account.py:165]: May want to artificially introduce some delay like `time.sleep()` to avoid sending too many requests to Gradescope at once

File: src/gradescopeapi/_classes/_account.py [hunk 4]
```diff
@@ -70,3 +77,124 @@ def get_courses(self) -> dict:
...
+            submission_links[submission_id] = aws_links
+        return submission_links
+
+    def get_assignment_submission(self, student_email, course_id, assignment_id):
+        # so far only accessible for teachers, not for students to get their own submission
+        # get_assignment_submission(token, student_id, course_id, assignment_id)
+        # -> link(s) to download file or actual files themselves
+        # fetch submission id
+        ASSIGNMENT_ENDPOINT = f"https://www.gradescope.com/courses/{course_id}/assignments/{assignment_id}"
+        ASSIGNMENT_SUBMISSIONS_ENDPOINT = f"{ASSIGNMENT_ENDPOINT}/review_grades"
```
Comment (calvinatian) [inline src/gradescopeapi/_classes/_account.py:178]:
```suggestion
        ASSIGNMENT_SUBMISSIONS_ENDPOINT = f"https://www.gradescope.com/courses/{course_id}/assignments/{assignment_id}/review_grades"
```

**Review by calvinatian — COMMENTED** [review]
Just a couple of questions and some nitpicks

File: src/gradescopeapi/_classes/_account.py [hunk 5]
```diff
@@ -70,3 +77,124 @@ def get_courses(self) -> dict:
...
+        if not course_id or not assignment_id:
+            raise Exception("One or more invalid parameters")
+        session = self.session
+        submissions_resp = check_page_auth(session, ASSIGNMENT_SUBMISSIONS_ENDPOINT)
+        submissions_soup = BeautifulSoup(submissions_resp.text, "html.parser")
+        submissions_a_tags = submissions_soup.select("td.table--primaryLink a")
+        submission_ids = [
+            a_tag.attrs.get("href").split("/")[-1] for a_tag in submissions_a_tags
+        ]
```
Comment (KevZhe) [inline src/gradescopeapi/_classes/_account.py:163]: being nitpicky here, but this is a lil hard to follow without visual separation and/or comments

File: src/gradescopeapi/_classes/_account.py [hunk 6]
```diff
@@ -70,3 +77,124 @@ def get_courses(self) -> dict:
...
+                        'assignment_id': 'a_id',
+                        'name': 'a_name,
+                        'release_date': 'a_release_date',
+                        'due_date': 'a_due_date',
+                        'late_due_date': 'a_late_due_date',
+                        'submissions_status': 'a_status',
+                        'grade': 'a_grade',
+                        'max_grade': 'a_max_grade'
+                   }
+                ]
```
Comment (KevZhe) [inline src/gradescopeapi/_classes/_account.py:99]: i think, to keep consistent with how Calvin did extensions and I did courses, the assignment attributes could be turned into a dataclass implemented in ```assignment.py```.  Inside the helper functions that scrape assignment info then, you could create these objects.

File: src/gradescopeapi/_classes/_assignment_helpers.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        # TODO: how should we handle errors so that our API can read them?
+        error_msg = [*json.loads(submissions_resp.text).values()][0]
+        if error_msg == "You are not authorized to access this page.":
+            raise Exception("You are not authorized to access this page.")
+        elif error_msg == "You must be logged in to access this page.":
+            raise Exception("You must be logged in to access this page.")
+    elif submissions_resp.status_code == requests.codes.not_found:
+        raise Exception("Page not Found")
+    elif submissions_resp.status_code == requests.codes.ok:
+        return submissions_resp
```
Comment (KevZhe) [inline src/gradescopeapi/_classes/_assignment_helpers.py:23]: nice, i think this would be useful as a general-purpose function for exception handling in other features

File: src/gradescopeapi/_classes/_account.py (hunk 3 above)
Comment (SushieeK) [reply src/gradescopeapi/_classes/_account.py:165]: introducing a delay such as time.sleep() could avoid sending too many requests to Gradescope in a short period.

**Review by SushieeK — CHANGES_REQUESTED** [review]

**Review by jmargaret2 — APPROVED** [review]

File: src/gradescopeapi/_classes/_account.py (hunk 3 above)
Comment (SushieeK) [reply src/gradescopeapi/_classes/_account.py:153]: Looks perfect after the changes.

**Review by KevZhe — APPROVED** [review]
lgtm

**Review by SushieeK — APPROVED** [review]
LGTM
